src/Storages/MergeTree/calculateBlockNumbersForRestoring.cpp

```diff
@@ -0,0 +1,482 @@
+                const auto & right = mutations[right_index];
+                return left.number < right.number;
+            });
+
+        apply_permutation(mutations, mutation_names_in_backup, std::move(permutation));
+    }
+
+    struct PartMutations
+    {
+        /// Index of the first mutation this part is applied to or should be applied to.
+        ///
+        /// For example, let's consider a MergeTree with six parts "all_1_2_1_8", "all_3_3_0", "all_4_4_0_5", "all_6_6_0_8", "all_7_7_0", "all_9_9_0"
+        /// and two mutations "mutation_5.txt", "mutation_8.txt". For those parts `first_mutation` and `num_applied_mutations` will be calculated as:
+        /// | part_name   | first_mutation | num_applied_mutations |
+        /// | all_1_2_1_8 | 0 (mutation_5) |                     2 |
+        /// | all_3_3_0   | 0              |                     0 |
+        /// | all_4_4_0_5 | 0              |                     1 |
+        /// | all_6_6_0_8 | 1 (mutation_8) |                     1 |
+        /// | all_7_7_0   | 1              |                     0 |
+        /// | all_9_9_0   | 2 (none)       |                     0 |
+        ///
+        size_t first_mutation = 0;
+
+        /// Number of mutations this part is already applied to.
+        size_t num_applied_mutations = 0;
+    };
+
+    /// Find corresponding mutations for each part.
+    std::vector<PartMutations>
+    findMutationsForEachPart(const std::vector<MergeTreePartInfo> & parts, const std::vector<MutationInfoFromBackup> & mutations)
+    {
+        std::vector<PartMutations> res;
+        res.resize(parts.size());
+
+        if (mutations.empty())
+        {
+            /// No mutations.
+        }
+        else if (mutations.front().block_number)
+        {
+            /// Mutations are specified with a single `block_number` per a mutation.
+
+            /// Make a map of mutations where mutations are ordered by their block numbers.
+            /// This map needs to be ordered because we use upper_bound() below.
+            std::map<Int64, size_t> mutations_by_block_number;
+
+            for (size_t i = 0; i != mutations.size(); ++i)
+            {
+                const auto & mutation = mutations[i];
+
+                /// MergeTree and ReplicatedMergeTree mutations can't be mixed.
+                if (!mutation.block_number || mutation.block_numbers)
+                    throw Exception(ErrorCodes::CANNOT_RESTORE_TABLE, "Mutations specify incorrect block numbers");
+                Int64 block_number = *mutation.block_number;
+
+                /// The block number specified in a mutation must increase with following mutations.
+                if (i > 0)
+                {
+                    Int64 previous_block_number = *mutations[i - 1].block_number;
+                    if (block_number <= previous_block_number)
+                        throw Exception(ErrorCodes::CANNOT_RESTORE_TABLE, "Mutations specify incorrect block numbers");
```

Comment:
This situation is possible because a backup can be damaged or written in a wrong way. I've changed the error message to make it more clear, see https://github.com/ClickHouse/ClickHouse/pull/50230/commits/d701dad2e70800f64c06361f8dcced8a78f5f5b1#diff-f56a47794b5c9778b9c556395f387a619390917a467a67be341242dda79b5a77R63

Reply:
Same, it still makes sense to use LOGICAL_ERROR code

Reply:
`LOGICAL_ERROR` means the situation is impossible from the logic of the code itself. If some files are corrupted this is not a logical error.

Reply:
Yes, and if files are corrupted, then we probably have a bug in logic that writes these files. The situation with broken files is impossible in the CI. It's also nearly impossible in production, and even if we get a broken backup somehow, it requires attention. So, I would like to have assertions for these errors in CI and alerts for these errors in production.

Reply:
How about `CORRUPTED_DATA`? It seems a pretty strong error code.

Reply:
Sounds good, we have alerts for this error code. It would be nice to catch it in the CI too. For example, we can add `chassert(false)` before throwing these exceptions. Or we can grep server logs for this error code after running tests like we grep "the specified key doesn't exist" in Stress tests

Reply:
I don't want to always write `chassert(false)` any time I throw a `CORRUPTED_DATA` error.
Perhaps this can be applied to all `CORRUPTED_DATA` errors, not just ones from this PR?
Let's modify the stress tests' logic, I'll do that in a separate PR, ok?